src/coreclr/nativeaot/System.Private.CoreLib/src/System/InvokeUtils.cs

```diff
@@ -134,87 +134,301 @@ private static Exception ConvertOrWidenPrimitivesEnumsAndPointersIfPossible(obje
+                    ushort ushortValue;
+                    switch (srcElementType)
+                    {
+                        case EETypeElementType.UInt16:
+                            ushortValue = (ushort)srcObject;
+                            break;
+                        case EETypeElementType.Char:
+                            ushortValue = (ushort)(char)srcObject;
+                            break;
+                        case EETypeElementType.Int32:
+                            ushortValue = (ushort)(int)srcObject;
+                            break;
+                        case EETypeElementType.UInt32:
+                            ushortValue = (ushort)(uint)srcObject;
+                            break;
+                        case EETypeElementType.Int64:
+                            ushortValue = (ushort)(long)srcObject;
+                            break;
+                        case EETypeElementType.UInt64:
+                            ushortValue = (ushort)(ulong)srcObject;
+                            break;
+                        case EETypeElementType.Single:
+                            ushortValue = (ushort)(float)srcObject;
+                            break;
+                        case EETypeElementType.Double:
+                            ushortValue = (ushort)(double)srcObject;
+                            break;
+                        default:
+                            goto Failure;
+                    }
                     dstObject = dstEEType->IsEnum ? Enum.ToObject(dstEEType, ushortValue) : ushortValue;
                     break;
 
+                // Int32 (W = I4, I8, R4, R8)
+                case EETypeElementType.Int32:
+                    int intValue;
+                    switch (srcElementType)
+                    {
+                        case EETypeElementType.Int32:
+                            intValue = (int)srcObject;
+                            break;
+                        case EETypeElementType.Int64:
+                            intValue = (int)(long)srcObject;
+                            break;
+                        case EETypeElementType.Single:
+                            intValue = (int)(float)srcObject;
+                            break;
+                        case EETypeElementType.Double:
+                            intValue = (int)(double)srcObject;
+                            break;
+                        default:
+                            goto Failure;
+                    }
+                    dstObject = dstEEType->IsEnum ? Enum.ToObject(dstEEType, intValue) : intValue;
+                    break;
+
+                // UInt32 (W = U4, I8, R4, R8)
                 case EETypeElementType.UInt32:
-                    uint uintValue = Convert.ToUInt32(srcObject);
+                    uint uintValue;
+                    switch (srcElementType)
+                    {
+                        case EETypeElementType.UInt32:
```

Comment:
Isn't this backwards? The method name says that it does widening, but this seems to implement narrowing.

Reply:
I... Uh... May have interpreted that type matrix backwards, yes 😅

Reply:
Fixed in 488e251bbbc22d5405008055700519e993967791.